core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java

```diff
@@ -620,6 +621,10 @@ private synchronized void updateSplitQueueSpace()
     private synchronized void processTaskUpdate(TaskInfo newValue, List<SplitAssignment> splitAssignments)
     {
         updateTaskInfo(newValue);
+        if (taskStatusFetcher.getTaskStatus().getState() == INITIALIZING) {
+            pendingRequestsCounter.incrementAndGet();
+            return;
```

Comment:
Add a comment why you return here. Is it needed?


Additionally, you need to schedule update when task transitions from `INITIALIZE` to `RUNNING`.

You should add a test that query won't deadlock, e.g:
1. Task is scheduled in `INITIALIZE` state
2. All splits (single split) is scheduled in `HttpRemoteTask`
3. Task transitions to `RUNNING` state
4. Expected: `HttpRemoteTask` will schedule task update on transition from `INITIALIZE -> RUNNING`

Reply:
I believe incrementing  `pendingRequestsCounter.incrementAndGet();` causes a reschedule of the update and will spin the update until the state is transitioned away. (The call site for this function has that logic).

Reply:
> pendingRequestsCounter.incrementAndGet();

It won't cause reschedule on it's own. You need to call `triggerUpdate`.

> and will spin the update until the state is transitioned away

We shouldn't spin `TaskUpdateRequest` actively until stage goes away, because `TaskUpdateRequest` (and `TaskInfo` responses are expensive). The proper way to do it is to listen for state changes from `taskStatusFetcher` and act upon these (e.g. call `triggerUpdate`)

BTW: is it OK to send splits while task is initializing? If so, maybe you don't need special handling in `HttpRemoteTask`

Reply:
Is the `taskStatusFetcher` always polling? And can used to block query execution until state ~= Initializing?

Reply:
I think I found a way to block stage execution until initialized is complete. I can probably remove this. But our latency would be limited by the latency of `ContinuousTaskStatusFetcher`?

Reply:
Is it a problem if a task fails to initialize after accepted splits?

Reply:
> Is the taskStatusFetcher always polling? And can used to block query execution until state ~= Initializing?

Yes. It's using long pulling.

> Is it a problem if a task fails to initialize after accepting splits?

Maybe not, but I don't know exactly how dynamic catalogs work